chain/chain/src/chain.rs

```diff
@@ -4132,29 +4136,37 @@ impl Chain {
         Ok(headers)
     }
 
-    /// Returns a vector of chunk headers, each of which corresponds to the previous chunk of
-    /// a chunk in the block after `prev_block`
+    /// Returns a vector of chunk headers, and vector of chunk endorsement signatures each of which corresponds
+    /// to the chunk in the `prev_block`
     /// This function is important when the block after `prev_block` has different number of chunks
-    /// from `prev_block`.
+    /// from `prev_block` in cases of resharding.
     /// In block production and processing, often we need to get the previous chunks of chunks
     /// in the current block, this function provides a way to do so while handling sharding changes
     /// correctly.
     /// For example, if `prev_block` has two shards 0, 1 and the block after `prev_block` will have
     /// 4 shards 0, 1, 2, 3, 0 and 1 split from shard 0 and 2 and 3 split from shard 1.
-    /// `get_prev_chunks(runtime_adapter, prev_block)` will return
+    /// `get_prev_chunk_headers_and_chunk_endorsements(epoch_manager, prev_block)` will return
     /// `[prev_block.chunks()[0], prev_block.chunks()[0], prev_block.chunks()[1], prev_block.chunks()[1]]`
-    pub fn get_prev_chunk_headers(
+    /// and the corresponding chunk endorsements.
+    pub fn get_prev_chunk_headers_and_chunk_endorsements(
         epoch_manager: &dyn EpochManagerAdapter,
         prev_block: &Block,
-    ) -> Result<Vec<ShardChunkHeader>, Error> {
+    ) -> Result<(Vec<ShardChunkHeader>, Vec<ChunkEndorsementSignatures>), Error> {
```

Comment:
I wonder if as a future refactoring idea, we should have

```
struct ChunkIncludedInBlock {
  header: ChunkHeader,
  height_included: u64,
  endorsement_signatures: Vec<ChunkEndorsementSignatures>
}
```

and BlockBody would have `Vec<ChunkIncludedInBlock>`.

This would simplify everywhere that we need to have two parallel vectors. But the height_included might be a more involved change, I'm not sure.

Reply:
In a follow up PR I can take a look at how we can simplify this logic. I agree, I wasn't too happy adding in the signatures here randomly.

Reply:
This PR should undo the complications caused here... https://github.com/near/nearcore/pull/10488